### PR TITLE
Header script minimization

### PR DESCRIPTION
**Pull request by crhallberg**
Automatically takes all the files requested and minimizes them into a single file. Skips files with conditionals (IE<10) for separate requests. Adds inline scripts last.

In order to use, you must create a cache folder: themes/{activeTheme}/js/concat.

- [x] Needs configuration
- [X] CSS support
- [x] Refactor for readability
- [x] Test how well this plays with less in theme.config.php
- [x] Determine best cache location
-  ~~Does not play well with less files linked and compiled from theme.config.php~~
- [x] Prevent external files from being packaged
- [x] Adjust .gitignore to account for cache
- [x] Investigate minification by group
- [x] Investigate bootstrap expansion phenomenon
- [x] Separate configuration for css and js
- [x] Change httpd-vufind.conf to reflect new path
- [x] Determine new cache location
- [ ] Close https://vufind.org/jira/browse/VUFIND-870 when done

### CODE REVIEW

**Comment by demiankatz** [comment]
More things to think about:

1.) This toString() method is now huge and significantly complicated -- I'd consider refactoring with support methods just to make it more readable. Perhaps some support methods could be moved to a Trait if that enables sharing of logic between the JS and CSS methods (once both exist).

2.) How do we want to handle conditional scripts? For example, if there's an IE-specific script, it looks like this current method will just go ahead and load it regardless. Do we need to sort scripts by conditions and have multiple hashed groupings? Do we want to say that this mechanism breaks conditional loading? Is there any other option?

3.) When we make this configurable, do we want to tie it to the environment (development vs. production) and/or a config.ini setting? If we go the config.ini route (which is almost certainly a good idea -- I don't think we want to force this option on anyone) we should be careful about how we inject the setting into the view helpers. Ideally I would like to keep any configuration accesses as high-level as possible (i.e. in the factories) so that the lower-level code is more reusable (no hard-coded coupling to specific configuration layouts). In theory I think the VuFindTheme module might eventually be a candidate for separation into its own project (though that's not a really high priority, as I think responsive design has made some of its capabilities less compelling than they used to be).

**Comment by crhallberg** [comment]
1.) It is complicated, looking for ways to refactor now.
2.) When a script is conditional, it is not included in the bundle and is pushed forward as normal with the conditionals for the browser to respond to.
3.) We'll talk about this once the code is cleaner.

**Comment by crhallberg** [comment]
Is it good practice to use references for get/set methods on objects? In set methods, it makes the return much easier, would it be significantly quicker for get?

Referencing: https://github.com/vufind-org/vufind/pull/757/files#diff-d338442e6dbc0dbc8c6988cfb3e21519R165

**Comment by demiankatz** [comment]
A thought on configurability -- what if we add a config.ini setting that's something like:

<pre>
asset_pipeline = *|production|development|false

* = always on
production = only on when APPLICATION_ENV == 'production'
development = only on when APPLICATION_ENV == 'development'
false = always off
</pre>

Rather than specifically implementing 'production' and 'development' cases, it could perhaps be done as something like `in_array(APPLICATION_ENV, array_map('trim', explode(',', $setting)))` to support a comma-separated list of arbitrary environment settings (in case that's of future value).

Not sure if there's any value in separate settings for JS vs. CSS or if we should just do one setting for both mechanisms.

**Comment by demiankatz** [comment]
Coming along nicely. Some suggestions:

1.) ConcatTrait has comments copied from a different trait. They should be adjusted, and the `@package

**Comment by demiankatz** [comment]
Coming along nicely. Some suggestions:

1.) ConcatTrait has comments copied from a different trait. They should be adjusted to actually describe the trait, and the `@package` should be updated to match the other files in the folder containing the file.

2.) It might be useful to include in the ConcatTrait comments a list of required methods that classes incorporating the trait must implement.

3.) In ConcatTrait, the $pipelineConfig property should have a `@var` comment block above it (so PHPDoc generates properly).

4.) Before you document it, though, I think it might be a good idea to replace the $pipelineConfig property with a simple boolean $usePipeline. This allows you to move all of the VuFind-specific logic involving config.ini and APPLICATION_ENV up out of the ConcatTrait and view helpers and into the factory. Create a static `usePipeline($sm)` method in the factory, and do the configuration loading/parsing there, reducing it down to a boolean. Then pass that boolean into the view helpers. I think that makes everything a bit cleaner.

**Comment by demiankatz** [comment]
Some minor TODOs:

1.) I don't claim that it makes sense, but the word is spelled "concatenate" -- there are some instances of "concatinate" that should be corrected.

2.) What do you think about creating a list of required methods in a comment on the ConcatTrait? If it's too much work, it's certainly not necessary -- but bringing it up again in case you missed my original comment about it.

**Comment by demiankatz** [comment]
One other suggestion: creating a function that takes a long list of reference parameters makes for fairly confusing code. Might it be easier to actually introduce some new properties in the ConcatTrait? I think this might improve readability a bit....

**Comment by crhallberg** [comment]
Is it better to put files in themes/CUSTOM/concat? Each theme will need its own. Or we can make a root theme concat folder and store files with a prefix of their current theme: bootstrap3hashashash.min.css.

**Comment by demiankatz** [comment]
Some more comments:

1.) I wonder if isResourceOtherItem() could be renamed to something more clear, like isEligibleForConcat() or isNonConcatItem() or shouldBeMerged() or isExcludedFromConcat(). If the real purpose of this method is to identify which things should not be concatenated, I think we should express that more directly in the method name. Asking whether it's an "other item" doesn't really say what an "other item" might be. It might make sense to rename the otherItems property to match the revised method name, if you agree that a change makes sense.

2.) In your toString() catch block, should you `return '';` so that at least some kind of string is being returned from the method? (Not sure if this matters, but seems like it might be good form).

3.) The comment above $this->filterItems() in ConcatTrait says "Returned by reference," which is no longer true. Should this instead say something like "Initialize class properties related to concatenation of resources"? Is it possible there's a better method name for this, like initItemsForConcatenation()?

4.) Would it make sense for filterItems() to return a boolean (true if items were found, false if they were not) so that you could simplify the top of your toString method to something like:

<pre>
            if (!$this->usePipeline || !$this->filterItems()) {
                return parent::toString($indent);
            }
</pre>
?

5.) Would it be safer for filterItems() to reset all properties to defaults at the top, so if the object is reused it doesn't have a weird state in it? (This possibility is, of course, the main disadvantage of using properties... another possible strategy would be to return a single data structure like an associative array containing all of the elements, instead of using properties... but that has its own set of annoyances).

6.) Is there really a need for the keyLimit property? It appears that you could just repeat your `foreach ($this as $key => $item) {` loop later on and get the same effect. (Apologies if I'm missing something). Similarly, do you actually need to store the concatTemplate as a property, or can you simply use concatIndex and manipulate the object at output time in outputInOrder? The fewer properties we have being used across multiple functions, the easier this will all be to understand.

7.) You have some comments that refer to "sheet object" in ConcatTrait; I think we probably want more generic language here.

8.) The $useCdata stuff in ConcatTrait appears to be JS-specific (at least, it seems to come from the ZF HeadScript helper but not be present in HeadLink). Is this a problem? (It looks like, in practice, it may not matter because the parameters being passed simply do not exist in HeadLink... but at very least it's potentially confusing/misleading).

9.) In the factory's getPipelineConfig method, you seem to be checking for a false value in two different places. I think you can eliminate the if/else in lines 60-62 because it's already covered by line 55.

10.) In your comment listing methods required by the ConcatTrait, you should also add a note about the fileType property, which is another important prerequisite.

11.) Regarding concatenation file location, I think I like the root theme as the location because it minimizes the number of locations that Apache is allowed to write to. I don't think you necessarily have to prefix the filename with the theme name, because the hash of filenames/modification dates should create non-colliding keys already.

Whew -- sorry for the lengthy list of comments... but since this is important low-level functionality, I want to be sure it's as clean and easy to understand as we can possibly make it.

**Comment by crhallberg** [comment]
1.-5., 7., 9.-10.) Done.

6.) Yes, order is very important because of dependency. In order to maintain order, I use for. If I use foreach, it'll skip some numbers. I could use a clever nested loop, but I thought this was simpler, if slightly messier.

8.) c-data is fine as far as I know, it should only apply to JS items. We could add more specific functions in HeadLink and HeadScript if you'd rather do that. It would duplicate more logic than it eliminates I believe.

11.) I'll work on it.

**Comment by crhallberg** [comment]
Less files added through theme.config.php are still included individually. Since the less files are compressed, can be `@include`d all together, and the practice of including less in theme.config.php is a developer convenience, I don't think this will pose major problems.

**Comment by demiankatz** [comment]
Some new food for thought:

- I noticed that the default bootprint3 CSS actually gets bigger after we minify it! The compiled.css is 183688 bytes, but the minified version in the cache grows to 259690 bytes. Do we want a special case that skips the cache if there's only one eligible file? Also, I note that we only minify when media="all" but I wonder if it would be better to sort the stylesheets by media type and potentially create a separate minified file for each grouping. (That may not be worth the effort given the minimum or even negative gains we're currently seeing... but in theory it seems like a more robust approach).

- It appears that offsite scripts get absorbed into the minified file -- for example, try turning on AddThis support and you will see an example. Is this desirable? It seems like it may be problematic, especially if those scripts change (since I don't think the file time check is necessarily going to work right for URLs wrapped by PHP file functions). Should we make a special exception for these?

- If we make more exceptions for things like offsite scripts, do we need to be more concerned about accurately retaining script order? Is it too risky to always concatenate down to a single file? Would it be better to concatenate adjacent groups of files and document the fact that users are responsible for optimizing behavior by shifting around their list of include statements?

**Comment by demiankatz** [comment]
I've just learned that Traits can implement abstract methods to impose requirements on consuming classes. See:

http://php.net/manual/en/language.oop5.traits.php#language.oop5.traits.abstract

This means that we can formalize the current comment about required methods into a set of abstract declarations that will actually enforce the requirement. Would you like to work on that when you have a chance, or should I?

**Comment by crhallberg** [comment]
Without any feedback from the dev call, I believe this is ready to go!

**Comment by EreMaijala** [comment]
1.) Needing write access to the theme directory is unfortunate and it seems to me it wouldn't play well with our structure where we have multiple VuFind instances with almost everything but a local "Custom" theme symlinked to a central directory. Would it be possible to put the minified files in the local cache directory which is already required to be Apache-writable and use aliases or rewrites in Apache to access them?

2.) A nitpick: this comment refers to css, but it's for js too:

`// Locate/create concatenated css file` 

3.) Is there somewhere a check that a concatenated css file doesn't become too large for IE < 10 to handle or is it practically impossible to happen? IE 9 and below have a limit of 4095 descriptors, and the rest is silently dropped (we actually have to split our less-compiled css file for IE since our rule count exceeds IE's limit, see https://github.com/NatLibFi/NDL-VuFind2/blob/master/module/Finna/src/Finna/View/Helper/Root/HeadLink.php#L153).

**Comment by demiankatz** [comment]
Thanks for the input, Ere! Comments on your comments:

1.) You make a good point that using the cache directory might be practical here, since including the theme name in the URL is not necessary for uniqueness due to the way the filenames are hashed. The main disadvantage to this approach would be the need for everyone to update their httpd-vufind.conf with a new block for the cache rewrite... but I suppose that's no more intrusive than requiring them to make an arbitrary directory apache-writable, and it's likely to be less fragile in the long term. @crhallberg, what do you think? On a semi-related note, I wonder if the existing LESS dynamic compilation could be adjusted to utilize the same directory to simplify management of that area as well.

2.) Thanks -- I'll let Chris fix that.

3.) I was not aware of this limitation. As we discussed earlier in this thread, the CSS minification seems to be of limited value as it is (in at least some cases, it makes the files bigger rather than smaller). This potential IE problem seems like another argument against minifying CSS by default and for allowing the user to independently configure which types of resources get minified.

**Comment by crhallberg** [comment]
1.) This method was designed to avoid a change to `httpd-vufind.conf` but if that's not a big deal, we could easily move the cache path. We'll have to come up with a better URL so it doesn't say `local/cache` in it. This is trivial, of course.
2.) Fixed.
3.) I also was unaware of this limitation. I'll expand configuration.

**Comment by demiankatz** [comment]
If we do decide to directly expose the cache through httpd-vufind.conf, we will need to make sure that the install.php script is capable of correctly updating the path to point to the right local settings directory even if the user puts this in an unusual place. (It may be covered by the default string-replace logic, but it's worth double-checking). Probably not too big of a deal, since we already have to update the file to set the VUFIND_LOCAL_DIR environment variable, and this is just a variation of that process.

However, just for the sake of exploring options, another possibility might be to create a new controller for fetching the cached files. This would avoid the need to modify httpd-vufind.conf at all, since we'd just access files using something like `http://server/vufind/Asset/Cache?key=abcd`. The very big and significant downside to this approach, of course, is the fact that every asset load would require us to fill up the full Zend Framework, which introduces significant overhead and would probably offset at least some of the benefit of using the pipeline in the first place.

**Comment by EreMaijala** [comment]
I thought about the option of using a controller to fetch the files, but it's so significantly more heavy that I wouldn't want to do it. Additionally the controller would need to handle If-Modified-Since and ETag request headers separately from Apache, which would mean Apache config couldn't be used to setup caching settings.

**Comment by demiankatz** [comment]
Yeah, you're right, it's not worth the effort and overhead. Let's see how it goes if we point into the local directory. @crhallberg, let me know if you need any help brainstorming the approach to this (not sure without looking at it how straightforward it will be).

**Comment by EreMaijala** [comment]
Just one thing: I'd add a new folder under the cache directory to make sure unrelated files are not http-accessible and make the cache manager (VuFind\Cache\Manager) auto-create it like any other cache subdirectory so that everything still works after cache is nuked. This also avoids any additional logic in installer.

**Comment by crhallberg** [comment]
@EreMaijala does a `/local/cache/public` suit your needs and system?

**Comment by EreMaijala** [comment]
Yes, thanks.

**Comment by EreMaijala** [comment]
The directory "/stage/local/vufind/local/cache/public/" in httpd-vufind.conf should be something more generic, right?

**Comment by crhallberg** [comment]
Good catch, I'm not sure what happened here, I was sure I fixed this directory.

**Comment by crhallberg** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;UDTZ4G5X&#x2F;40-header-min-script-cache-dir-feedback"> Header Min Script (cache dir feedback)</a>

File: module/VuFindTheme/src/VuFindTheme/View/Helper/ConcatTrait.php [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+        $output = [];
+        foreach ($this->groups as $group) {
+            if (isset($group['other'])) {
+                $output[] = $this->itemToString(
+                    $group['item'], $indent, $escapeStart, $escapeEnd
+                );
+            } else {
+                $path = $this->getConcatenatedFilePath($group);
+                $item = $this->setResourceFilePath($group['items'][0], $path);
+                $output[] = parent::itemToString(
```
Comment (demiankatz) [inline module/VuFindTheme/src/VuFindTheme/View/Helper/ConcatTrait.php]: Why are we using parent::itemToString here and $this->itemToString above in line 265? If this is a meaningful distinction, we should comment it. If not, we should switch to something more consistent.

**Comment by demiankatz** [comment]
All done. @EreMaijala, please let us know if you have any trouble with this. The CSS rule limit is not addressed in this code, but I'd recommend just turning on the pipeline for JS and ignoring the CSS part -- there aren't big gains to be had there anyway.

**Pull request merged**